includes/Core/Assets/Assets.php

```diff
@@ -230,6 +230,41 @@ private function register_assets() {
 		foreach ( $assets as $asset ) {
 			$asset->register();
 		}
+
+		$this->add_amp_dev_mode_attributes( $assets );
+	}
+
+	/**
+	 * Add data-ampdevmode attributes to assets.
+	 *
+	 * @todo What about dependencies?
+	 *
+	 * @param Asset[] $assets Assets.
+	 */
+	private function add_amp_dev_mode_attributes( $assets ) {
+		add_filter(
+			'script_loader_tag',
+			function ( $tag, $handle ) use ( $assets ) {
+				if ( $this->context->is_amp() && isset( $assets[ $handle ] ) && $assets[ $handle ] instanceof Script ) {
```

Comment:
This apparently needs to look to also look to see if the `$handle` is a dependency of any `$assets` as well.

Reply:
We can probably add a `get_dependency_handles()` method to `Google\Site_Kit\Core\Assets\Asset`, which we can then leverage here to check that.

Reply:
Out of curiosity, why would the amp plugin not just add the attribute to all style and script tags? Why only whitelist Site Kit assets and dependencies? If needed, why wouldn't the amp plugin add the attributes and let other plugins hook in to whitelist its assets rather than each plugin reimplementing this itself?

Reply:
The attribute should only be added to JS that is related to functionality not relevant to the frontend (and which thus would not actually break AMP compatibility of the frontend, as long as you're logged-out). That's why it needs to be conservative, any plugin that has JS-powered admin bar functionality needs to explicitly declare that.

Regarding addition of the attributes, that's a good point. @westonruter Maybe the AMP plugin could include a central solution for adding the attribute that scripts/stylesheets could leverage via e.g. a `wp_*_add_data( ... )`?

Reply:
Note that any script or stylesheet that has `admin-bar` as a dependency will automatically get the dev mode attribute. The issue is with scripts/styles that do not have this explicit admin bar linkage, such as dependencies on general libraries. 

> Maybe the AMP plugin could include a central solution for adding the attribute that scripts/stylesheets could leverage via e.g. a `wp_*_add_data( ... )`?

I'm not sure I understand. Are you saying you could do something like:

```php
wp_script_add_data( 'lodash', 'ampdevmode', true );
```

And that this could be used as a dev mode signal equivalent to dependence on the `admin-bar` script? Then the AMP plugin could include a `script_loader_tag` filter which adds the `data-ampdevmode` attribute automatically to any script that has the `ampdevmode` data or which depends on a script that has this attribute?

This could also be used to automatically add the comment for inline scripts as well. Whereas right now we are manually adding `/* googlesitekit */` to the inline scripts, there could be a `wp_print_scripts` action which loops over all scripts prior to printing and any script which contains the `ampdevmode` data could also then do:

```php
wp_add_inline_script( $handle, '/*ampdevmode*/' );
```

And this pattern could be automatically added to the list of `amp_dev_mode_element_xpaths`.

Does this sound right?

Reply:
@westonruter Yes, that would be great - really like including inline script support as well.

Reply:
Work in progress: https://github.com/ampproject/amp-wp/pull/4001